Formula/q.rb

```diff
@@ -17,14 +15,27 @@ class Q < Formula
     sha256 cellar: :any_skip_relocation, x86_64_linux:   "bed14a331133ff96b85fa37e0729ca695bd273f78ee82e792185d137edf9917a"
   end
 
+  depends_on "pyoxidizer" => :build
+  depends_on "python@3.8" => :build
   depends_on "ronn" => :build
-  depends_on "python@3.9"
-  depends_on "six"
+  depends_on xcode: ["12.4", :build]
+
+  # Patch to allow proper ARM builds by using python3.9.
+  # Will be removed in the next version since patch is already in the master
+  patch :DATA
 
   def install
-    # broken symlink, fixed in next version
-    rm_f "bin/qtextasdata.py"
-    virtualenv_install_with_resources
+    arch_folder = if OS.linux?
+      "x86_64-unknown-linux-gnu"
+    elsif Hardware::CPU.intel?
+      "x86_64-apple-darwin"
+    else
+      "aarch64-apple-darwin"
+    end
+
+    system "pyoxidizer", "build", "--release", "--var", "PYTHON_VERSION", "3.9"
```

Comment:
Let's not hardcore this.
```suggestion
    xy = Language::Python.major_minor_version("python3")
    system "pyoxidizer", "build", "--release", "--var", "PYTHON_VERSION", xy
```

Reply:
As I described above, configuring pyoxidizer to use python 3.9 here is unrelated to the python of the runtime environment, it's a configuration meant to tell it which version of cpython needs to be used in order to create the standalone executable (after pyoxidizer is done, no python dependency is needed at all).

Reply:
in the next version of q, all the binaries that pyoxidizer issues will be based on python 3.9/3.10, so this parameter won't be needed anymore.

Reply:
Ok, but does it actually use a Python interpreter during the build? Does it need a Rust compiler?

Reply:
It doesn't actually use a python interpreter during the build, this is why i said it's a good catch :)

I've removed the python dependency entirely. See my next commit.

Also, q doesn't require a rust compiler as a dependency, this is managed internally in the pyoxidizer runtime (similar to the way it downloads the relevant CPython version during the build).

Reply:
> Also, q doesn't require a rust compiler as a dependency, this is managed internally in the pyoxidizer runtime (similar to the way it downloads the relevant CPython version during the build).

Well, this is why I asked -- we don't want it downloading its own CPython or Rust during the build. It should use the formulae instead.

Reply:
I see. 

This is how pyoxidizer works, and is a major part of its value - It takes a huge complication of matching the relevant cpython+rust needed for the platform that's being built (which include native libraries relevant for the platform), and generates a single standalone executable which does not depend on any of them at all (a fully native standalone executable). The amount of work that's being done by pyoxidizer is tremendous and cannot be replicated manually, definitely not for multiple architectures as is the case for a brew formula.

The Rust+CPython that are downloaded during build are just implementation details of the way pyoxidizer works. Essentially, pyoxidizer's entire value is based on that, and it's a very famous project which becomes the new standard for compiling python code into standalone native executables with no dependencies. It's also actually available itself as a brew Formula.

Can you please describe why this would be a problem? Those downloads during builds are essentially identical to any other 3rd party library that any other build tool would download during build (e.g some npm library during npm install, or python package downloaded from pip). Those could also include platform-specific logic that could fail during build. The way to make sure that everything works is the test part in the formula, which works well for all architectures required by brew.

Reply:
> The amount of work that's being done by pyoxidizer is tremendous and cannot be replicated manually, definitely not for multiple architectures as is the case for a brew formula.

Why doesn't adding `depends_on "rust"` and using the [`--system-rust`](https://pyoxidizer.readthedocs.io/en/stable/pyoxidizer_managing_projects.html#managed-rust-toolchain) flag work for Rust?

> Those downloads during builds are essentially identical to any other 3rd party library that any other build tool would download during build (e.g some npm library during npm install, or python package downloaded from pip).

Likening this to `npm` is actually worrying, because the penchant of `npm` packages to use (and often install!) pre-built binary executables/libraries is a constant source of headaches for us. As for `pip`: if you look at how we use `pip` in Homebrew, we only ever use it to install sdists, which are themselves built from source. (Hence why a number of Python-based formulae have a Rust dependency to build `cryptography` from source.)

> Can you please describe why this would be a problem?

For one thing, we [build everything from source in Homebrew/core](https://docs.brew.sh/Acceptable-Formulae#we-dont-like-binary-formulae), and that includes the tools used to build formulae. (We do make exceptions for formulae which need a version of themselves to bootstrap, but that doesn't apply here.) It helps to: 1) reduce duplication, 2) ensure compatibility on the systems we install things on, and 3) provide better (or, at least, more transparent) guarantees about the security of the binary objects that we use.

Reply:
<Sorry for the long post. English is not my native tongue, and i'm trying to be as understandable as possible, while maintaining a professional discussion and good vibe. @carlocab - I'm not trying in any way to rule-out or under-appreciate your concerns>

The rust part, and using `--system-rust`` is indeed an option, although a cumbersome one, since it would require managing the specific version/s of rust for which pyoxidizer is compatible with. I'd gladly go for it if you believe that would be better, although I believe this would result in a less-stable build process in the long-run. The reason for that is that pyoxidizer can decide to bump the version of rust their using internally for the build process, which could break things even though it's just an implementation detail of the tool.

About the part related to CPython, the comparison to npm was just an example (that i can definitely relate to). However, I don't view it as being similar in terms of stability, since the python distributions are being managed by pyoxidizer itself ([here](https://python-build-standalone.readthedocs.io/en/latest/)), and are used by it, contrary to the case of npm where it's a "wild west" situation.

One of the main advantages of pyoxidizer is to be able to choose the correct cpython for the platform at hand, including a build/ci process on their side that makes sure that the chosen cpython version matches the needs of the platform properly, including all other dependencies and native python extensions etc. I'm not sure that it's even possible to provide an external cpython installation which will work with pyoxidizer, other than just downloading the exact cpython tar.gz which pyoxidizer actually uses for the build. Essentially, the rust project generated and compiled by pyoxidizer links with the cpython release, which could be considered as a form of "build-from-source" - the rust project would not build properly if there's no match between the platform and the relevant cpython code.

Would removing that rust dependency by adding a direct dependency on rust and using `--system-rust` suffice in order to ease your mind about this? I'd really appreciate it if it would.

If not, then it would seem to me like more of a macro issue of stating that pyoxidizer is incompatible with brew's requirements, which would be in my view as a waste of a hugely beneficial tool that can take the industry to the next level with regard to python distribution. The fact that pyoxidizer's purpose is to generate a standalone executable which does not depend on anything stresses this further, because possible instabilities are delegated to the build process instead of the runtime.

Btw, pyoxidizer's build process and the resulting binaries (I use it for OSX/Linux/Windows, and all OSX variants are passing compile and test here in brew), seems to be hugely more robust and stable than all of the other similar solutions I've seen in the last years.

Reply:
I'm the maintainer of PyOxidizer and can provide technical assistance.

I think the crux of the concern in this issue is that Homebrew maintainers want the building of packages *utilizing* PyOxidizer at build time to have the build process be as self-contained and reliant on other Homebrew packages as possible. That means invocations of `pyoxidizer` (the CLI executable provided by PyOxidizer) would only consume files from the local filesystem.

This is a reasonable stance for packaging system maintainers to take.

Unfortunately, a hard line stances would effectively prohibit `pyoxidizer` from being utilized in Homebrew because - as @harelba correctly notes - `pyoxidizer` will - by default - download pinned/deterministic/cryptographically validated copies of Rust and Python at run-time (Rust is used for building the resulting binaries and Python is used at build time and is embedded in the resulting binary).

The Rust dependency is easy to circumvent: just pass `--system-rust` and `pyoxidizer` should happily consume the `cargo` executable in PATH (instead of using its own managed copy). In this world, Homebrew would add a PyOxidizer formula that depends on the version of Rust that PyOxidizer itself uses by default and formulae using `pyoxidizer` at build time would call `pyoxidizer --system-rust` to pick up the Homebrew managed Rust.

The Python dependency is less trivial to work around. In order to support cross-compilation and to customize the Python interpreter embedded in produced binaries, PyOxidizer consumes special _standalone_ Python distributions produced via https://github.com/indygreg/python-build-standalone. These Python distributions are downloaded at `pyoxidizer` run-time and used at both build time (for running Python packaging tools like `pip` for example) and at run-time (they are embedded in the resulting binary).

I see a few potential outcomes here:

1. Homebrew keeps things simple and allows `pyoxidizer` to download Python distributions at build time. The distributions are hosted as GitHub release artifacts and have their SHA-256 pinned. So behavior should be deterministic and reproducible given the same build of `pyoxidizer`.
2. Homebrew creates formulae to install the special Python distributions from https://github.com/indygreg/python-build-standalone. Homebrew adopts a policy that any formulae using `pyoxidizer` must consume the Homebrew-managed Python distributions. (PyOxidizer supports specifying the location of the Python distributions: https://pyoxidizer.readthedocs.io/en/pyoxidizer-0.18.0/pyoxidizer_config_type_python_distribution.html#starlark_pyoxidizer.PythonDistribution.__init__. 
3. Homebrew bans use of `pyoxidizer` at formulae build time. (But Homebrew could still provide a formula for PyOxidizer itself to enable people to easily install the tool.)

2 scares me a bit because PyOxidizer really wants specific Python distributions to be used and is only well tested against the artifacts referenced by PyOxidizer's source code. So if Homebrew introduces drift between what PyOxidizer uses by default and what Homebrew has installed, that will likely result in bugs filed against PyOxidizer, Homebrew, and individual packaged applications.

I think the important question is **does Homebrew want formulae providing applications/binaries embedding their own Python distribution?** If *no*, then `pyoxidizer` likely can't be used at formula build time, as `pyoxidizer` will produce binaries that have their own copy of `libpython` and the Python standard library. If *no*, application maintainers will have to coerce their Python applications into working like *traditional* Python applications (using a Python interpreter and standard library shared by multiple Python applications) in order to package it with Homebrew.

PyOxidizer is somewhat unconventional in much of what it does. That's a large part of its value proposition. So I'm not surprised to see some packaging hiccups related to its use. I hope this post provided some useful context.

Reply:
Thanks for the detailed write-up here, @indygreg. That certainly clarifies a lot, and I really appreciate how much you've distilled the important issues here.

In particular, I think this is indeed of central importance:

> **does Homebrew want formulae providing applications/binaries embedding their own Python distribution?**

I'm afraid the answer here is no. This sounds a lot to me like statically linking dependencies, and that's something we generally frown upon. (This is a pretty common stance among package managers.)

I think we'll have to go with the third option you present. We're happy to keep packaging pyoxidizer; it seems like a nice tool that many users would find useful. I don't think we want formulae we package to actually use it, however.

CC @Homebrew/core as an FYI re `pyoxidizer` dependencies, and in case you have additional thoughts here.

Reply:
By the way, @indygreg, for future reference: is there a way to instruct PyOxidizer to build the embedded Python from source, or would that need to be done separately with PyOxidizer directed to the specially built Python?

Reply:
PyOxidizer doesn't yet have a mechanism for building the embedded Python. The best you can do is for application config files (`pyoxidizer.bzl`) to point at custom path/URL (https://pyoxidizer.readthedocs.io/en/pyoxidizer-0.18.0/pyoxidizer_config_type_python_distribution.html#starlark_pyoxidizer.PythonDistribution.__init__).

There's probably room to introduce a global CLI argument or run-time behavior to define an alternate location for the default Python distributions.

The build system for https://github.com/indygreg/python-build-standalone is reasonably deterministic. I'm guessing downstream packagers of PyOxidizer not wanting the download-a-file-from-the-internet-at-runtime behavior will build their own distributions (or copy/vendor the ones I make), make them available via a package, then modify PyOxidizer's built-in default distributions to pick up the `.tar.zst` archives from a local filesystem path. https://github.com/indygreg/PyOxidizer/blob/main/pyoxidizer/src/python_distributions.rs#L64 is where those are defined.

The macOS side of the python-build-standalone build system isn't as deterministic as Linux. So if Homebrew proceeds with building their own Python distributions, please let me know so I can look at the CI and possibly shore up the auditing to ensure things are built properly (there's all kinds of subtle issues with macOS SDK versions, weak linking, etc that need to be _just right_ for binaries built with PyOxidizer to _just work_).

Reply:
> there's all kinds of subtle issues with macOS SDK versions, weak linking, etc that need to be _just right_ for binaries built with PyOxidizer to _just work_

I understand why you're concerned with such issues, but there's really very little need to worry about such things for binaries distributed via Homebrew. This is because we do almost the opposite of cross-compilation on macOS: we only build binaries on machines as similar to the target as possible. The typical case is that binaries built on, say, macOS 10.15 are only run on macOS 10.15. Occasionally, a user will have something installed that was built on an older macOS version, but experience shows that this is basically never a problem.

This makes all but the most subtle portability issues go away.